root/meta/race-12552/exec_dlopenrace.C

```diff
@@ -0,0 +1,47 @@
+#include <thread>
+#include "TClassTable.h"
+#include "TROOT.h"
+#include <dlfcn.h>
+#include <iostream>
+
+void repeatopen(int n)
+{
+   for(int i = 0; i < n; ++i) {
+     auto r = dlopen("./libUser.so", RTLD_LAZY);
+     if (!r) {
+        std::cerr << "Failed to open libUser.so at iteration " << i << "\n";
+        return;
+     }
+     // Waste some time
+     for(int j = 0; j < 10; ++j)
+       auto d = TClassTable::GetDictNorm("UserClass");
+     // std::cerr << "Seeing dict pointer: " << (void*)d << '\n';
+     dlclose(r);
+   }
+}
+
+void repeatlookup(int n)
+{
+   static DictFuncPtr_t value = nullptr;
+
+   for(int i = 0; i < n; ++i) {
+     auto r = TClassTable::GetDictNorm("UserClass");
+     if (!value && r)
+       value = r;
+     // std::cerr << "Value of dictionary pointer has from " << (void*)value << " to " << (void*)r << '\n';
+     if (r != nullptr && r != value)
```

Comment:
Why not simply `if (r != value)`?  Is it allowed to change back to `nullptr` even if we had a value before?

Reply:
`r == nullptr && r != value` happens and is 'expected' in this test in the case where `value` has already been assigned and `TClassTable::GetDictNorm` is called between the invocation of `dclose` and the invocation of `dlopen` (i.e. at a point in time where the library is unloaded).